src/routes/threlte/NewCircleTool.svelte

```diff
@@ -1,128 +1,152 @@
+
+	export function mouseMove(_event: Event, projected: { x: number; y: number }) {
 		// search through the existing points to see if we're close to one
 		// if we are, then we should snap to it
 
 		// TODO: in the future, we should also snap to the midpoints of lines
 		// and to the perimeters of circles and so on
 		// so these snap points do not necessarily correspond to actual points in the sketch
-		let snappedTo
-		for (let geom of $currentlyMousedOver) {
+		let snappedTo = null
+		for (const geom of $currentlyMousedOver) {
+			// log("[currentlyMousedOver geom]", geom)
 			if (geom.type === 'point3D') {
-				let twoD = projectToPlane(new Vector3(geom.x, geom.y, geom.z))
-				let point = {
+				const twoD = projectToPlane(new Vector3(geom.x, geom.y, geom.z))
+				// log("[projectToPlane twoD]", twoD)
+				const point = {
 					twoD: { x: twoD.x, y: twoD.y },
 					threeD: { x: geom.x, y: geom.y, z: geom.z },
-					pointId: null
+					id: null
 				}
 				snappedTo = point
 			}
 			if (geom.type === 'point') {
-				let point = pointsById[geom.id]
-				snappedTo = { twoD: point.twoD, threeD: point.threeD, pointId: geom.id }
+				// log("[currentlyMousedOver geom is type point]", geom)
+				const point = pointsById[geom.id]
+				// oops! point.twoD etc does not exist here, we have:
+				// const example = {
+				// 	type: "point",
+				// 	id: "1"
+				// }
+				function querySnapPoint(id: string | null) {
+					const points = $snapPoints.filter((point) => id && point.id === id)
+					return points.length > 0 ? points[0] : false
+				}
+				// see if we can retrieve it? unlikely
+				// log("[querySnapPoint found point:]", querySnapPoint(point?.id!))
+				// have not seen a successful query yet! sort it out with an if:
+				if (point.twoD && point.threeD && geom.id)
+					snappedTo = { twoD: point.twoD, threeD: point.threeD, id: geom.id }
 				break // If there is a 2D point, prefer to use it rather than the 3D point
 			}
 		}
 
+		// if (snappedTo) log("[snappedTo]", snappedTo)
+
 		// only reset $snapPoints if something has changed
 		if (snappedTo) {
-			$snapPoints = [snappedTo]
+			// @ts-ignore todo rework snapping
+			$snapPoints = [snappedTo] // todo all these different point representations need work!
 		} else {
 			if ($snapPoints.length > 0) {
 				$snapPoints = []
 			}
 		}
 
 		if (centerPoint) {
-			let radius
-			if (snappedTo) {
-				let dx = snappedTo.twoD.x - centerPoint.twoD.x
-				let dy = snappedTo.twoD.y - centerPoint.twoD.y
-				radius = Math.hypot(dx, dy)
-			} else {
-				let dx = projected.x - centerPoint.twoD.x
-				let dy = projected.y - centerPoint.twoD.y
-				radius = Math.hypot(dx, dy)
+			function calcDeltas(
+				a: Vector2Like | Point2D | { x: number; y: number },
+				b: Vector2Like | undefined
+			) {
+				const dx = a.x - b?.x!
+				const dy = a.y - b?.y!
+				return Math.hypot(dx, dy)
 			}
+			const radius = snappedTo
+				? calcDeltas(snappedTo.twoD, centerPoint.twoD)
```

Comment:
ha! neat

Reply:
I like functions! and ternary operators! plus I like it when it's easier for my brain to understand :grin: